src/sybild/sybild.proto

```diff
@@ -0,0 +1,210 @@
+  repeated string distinct_by = 7;
+  string sort_by = 8; // field to sort by.
+  // todo: prune_by
+
+  string time_column = 9; // column to consider as the time column.
+  int64 time_bucket = 10; // time bucket size in seconds.
+
+  string weight_column = 11; // the column to interpret as the weight (for sampling).
+
+  QueryOp op = 12; // the operation to run.
+
+  repeated QueryFilter int_filters = 13; // filters on int columns.
+  repeated QueryFilter str_filters = 14; // filters on string columns.
+  repeated QueryFilter set_filters = 15; // filters on set columns.
+
+  HistogramOptions histogram_options = 16;
+
+  bool read_ingestion_log = 17;
+  // todo: query cache control?
+}
+
+// HistogramType describes the type of histogram.
+enum HistogramType {
+  HISTOGRAM_TYPE_UNKNOWN = 0;
+  NORMAL_HISTOGRAM = 1;
+  LOG_HISTOGRAM = 2;
+}
+
+// HistogramOptions
+message HistogramOptions {
+  HistogramType type = 1
+  int64 bucket_size = 2;
+}
+
+// QueryResult is the type of the query result.
+message QueryResult {
+  map<string, Histogram> histograms = 1;
+  int64 distinct = 2;
+
+  string group_by_key = 3;
+  int64 count = 4;
+  int64 samples = 5;
+}
+
+// Histogram describes a distribution of values.
+message Histogram {
+  double mean = 1;
+  int64 max = 2;
+  int64 min = 3;
+  int64 total_count = 4;
+
+  repeated int64 percentiles = 5;
+  map<int64, int64> int_buckets = 6;
+
+  int64 range_start = 7;
+  int64 range_end = 8;
+  double std_deviation = 9;
+}
+
+// ResultMap
+message ResultMap {
+  map<string, QueryResult> values = 1;
+}
+
+// SetField
+message SetField {
+  repeated string values = 1;
+}
+
+// Record is an individual record.
+message Record {
+  map<string, string> strs = 1;
+  map<string, int64> ints = 2;
+  map<string, SetField> sets = 3;
+}
+
+// QueryResults
+message QueryResults {
+  QueryResult cumulative = 1;
+  ResultMap results = 2;
+  map<int64, ResultMap> time_results = 3;
+  int64 matched_count = 4;
+  repeated QueryResult sorted = 5;
+  repeated Record  matched = 6;
+}
+
+// QueryResponse
+message QueryResponse {
+  QueryResults results = 1;
+}
+
+// ListTablesRequest
+message ListTablesRequest {
+}
+
+// ListTablesResponse
+message ListTablesResponse {
+  repeated string tables = 1;
+}
+
+// GetTableRequest
+message GetTableRequest {
+  string name = 1;
+}
+
+// Table
+message Table {
+  string name = 1;
+  repeated string str_columns = 2;
+  repeated string int_columns = 3;
+  repeated string set_columns = 4;
+  int64 count = 5;
+  int64 storage_size = 6;
+  int64 average_object_size = 7;
+}
+
+// TrimRequest queries and optionally removes data from a dataset.
+message TrimRequest {
+  string dataset = 1;
+  string time_column = 2;
+  int64 megabyte_limit = 3;
+  int64 time_limit = 4;
+  bool delete = 5; // if supplied, actually remove matching blocks.
+}
+
+// TrimResponse returns the number of blocks matched by the provided thresholds.
+message TrimResponse {
+  int64 matched_blocks = 1;
+}
```

Comment:
maybe also return the timestamp that was chosen if the trim happens based on a memory limit. f.e. "trim until dataset is 2gb" should return the timestamp that all data was deleted before

Reply:
:+1:  - added explanatory text.